File: front/src/lib/evolutions.ts
import { separerFichiers } from '$lib/fichiers';
import { perimetreHerite } from '$lib/perimetres';
import {
	motifInactif,
	SECTIONS_ORDRE,
	sectionPresente,
	type ConditionInactive,
	type EntiteDeclaree,
	type IdSection,
} from '$lib/entites/types';
/**
 * Le vocabulaire d'un **fil d'évolution** — écrit une fois pour les trois entités
 * qui en portent un : tickets, actualités, événements de calendrier.
 *
 * ## Pourquoi ce module (19/08/2026, signalé à l'écran)
 *
 * > *« Pourquoi mon commentaire sur un ticket a une icône de type relance et non
 * > commentaire ? »*
 *
 * Parce que la même notion était écrite **trois fois, avec trois valeurs** :
 *
 * | Où | `commentaire` | `reponse` | `etat` |
 * |---|---|---|---|
 * | le bouton qui ouvre le formulaire (`HistoriqueTicket`) | 💬 | — | — |
 * | le fil lui-même (`RubriqueHistorique`) | 📝 | 💬 | 🔄 |
 * | le flux d'activité (`api/app/routers/flux/tickets.py`) | 🔧 | 💬 | — |
 *
 * On cliquait donc « 💬 Commenter » et l'entrée s'affichait en 📝 — un mémo, que
 * l'utilisateur a lu comme une relance. Et 💬, l'icône qu'il attendait, était
 * prise par `reponse`.
 *
 * C'est le défaut de #415 (les statuts) et #413 (les champs), sur un troisième
 * objet : *« Chacune était cohérente avec elle-même ; c'est ce qui les rendait
 * invisibles à la relecture. »*
 *
 * ⚠️ **Le geste et son résultat doivent porter le MÊME signe.** C'est la règle
 * qui tranche ici : l'icône de l'entrée est celle du bouton qui l'a créée, pas
 * l'inverse. Un utilisateur ne relit pas une table, il reconnaît un dessin.
 *
 * ## Ce qui reste écrit deux fois, et pourquoi
 *
 * L'API a sa propre table (`flux/tickets.py`) : les contextes de build sont
 * `./api` et `./front`, rien de la racine n'entre dans les images — aucun fichier
 * ne peut être partagé (cf. la mémoire projet du 14/08/2026). Elle décrit
 * d'ailleurs un AUTRE rendu — une carte de flux « ticket mis à jour », pas une
 * entrée de fil — donc son 🔧 n'est pas forcément faux. L'écart est signalé, pas
 * corrigé en silence.
 */

/** Les trois types que porte une entrée de fil, côté serveur comme côté écran. */
export type TypeEvolution = 'commentaire' | 'etat' | 'reponse';

/**
 * Icône d'une entrée de fil.
 *
 * `commentaire` → 💬, **la même que le bouton « Commenter »**.
 * `reponse` → ↩️ : elle répond à quelque chose, ce que la bulle seule ne disait
 * pas — et la bulle revient à qui la mérite.
 * `etat` → 🔄, inchangée : une transition de workflow.
 */
export const EVOLUTION_ICONE: Record<string, string> = {
	commentaire: '\u{1F4AC}',
	reponse: '↩️',
	etat: '\u{1F504}',
};

/**
 * L'icône du type, ou celle du commentaire par défaut.
 *
 * ⚠️ Le repli est `commentaire` et non un point d'interrogation : un type inconnu
 * vient forcément d'une entrée de fil, et lui donner un signe d'erreur ferait
 * croire à un défaut de la donnée là où il n'y a qu'un type que l'écran ne
 * connaît pas encore.
 */
export function evolutionIcone(type: string | undefined | null): string {
	return EVOLUTION_ICONE[type ?? ''] ?? EVOLUTION_ICONE.commentaire;
}

/**
 * **La charge utile qu'un formulaire d'évolution émet** — le contrat entre
 * `EvolForm` et ceux qui la relaient à l'API.
 *
 * ## Pourquoi ce type existe (#529, 20/08/2026)
 *
 * Signalé à l'écran : *« j'ai créé une réponse au ticket en changeant le
 * périmètre, celui-ci n'a pas été pris en compte »*.
 *
 * `CarteTicket` proposait bien la section Périmètre, `EvolForm` la collectait et
 * l'émettait — et `tickets/+page.svelte` la **jetait** en recopiant la charge
 * utile champ par champ, à partir d'un type local qui l'ignorait. Ce type local
 * portait pourtant le commentaire *« même contrat que la fiche détail »*, ce qui
 * était faux : la fiche, elle, relaie la charge entière.
 *
 * 🔴 **Le défaut ne lève rien.** Le formulaire annonce l'enregistrement, le
 * serveur enregistre une évolution parfaitement valide, et seul le périmètre
 * affiché ensuite trahit la perte. C'est le profil d'erreur qu'aucun test
 * fonctionnel ne voit et qu'une relecture ne trouve pas — il faut comparer deux
 * fichiers distants de quatre cents lignes.
 *
 * ⚠️ Un champ ajouté ici doit l'être **aussi** dans le `dispatch` d'`EvolForm` et
 * dans le client d'API. `npm run lint:charge-utile` échoue si un relais oublie
 * un champ que le formulaire émet.
 */
export interface ChargeUtileEvolution {
	/** « Rédigé avec l'assistant IA » (#985) — seulement quand c'est vrai. */
	assiste_ia?: boolean;
	type: string;
	contenu?: string;
	nouveau_statut?: string;
	fichiers_urls?: string[];
	email_externe?: string;
	partager_whatsapp?: boolean;
	envoyer_syndic?: boolean;
	envoyer_cs?: boolean;
	/**  « M'envoyer une copie » — la 4e case de la Diffusion (31/08/2026). Un
	 *   CHOIX, jamais un envoi implicite : le formulaire annonçait trois
	 *   destinataires et en servait quatre. */
	envoyer_auteur?: boolean;
	/**  Le périmètre que l'entrée PRÉCISE — absent quand elle n'en parle pas, et
	 *   le serveur ne touche alors pas à celui de l'objet (#497). */
	perimetre_cible?: string[];
	/**  À qui l'on parle — émis quand la section Destinataires est affichée, et
	 *   appliqué par le serveur au conseil seul (une actualité, #1091). */
	public_cible?: string[];
	/**  Message interne : proposé seulement là où `avecInterne` est activé, donc
	 *   aujourd'hui la seule fiche d'un ticket. */
	interne?: boolean;
	/**  🔴 LES OPTIONS DE PUBLICATION D'UN TICKET (05/09/2026) — corrigées depuis
	 *   un commentaire, comme sur une actualité : le formulaire montre le dernier
	 *   état, ce qu'on enregistre devient l'état.
	 *
	 *   ⚠️ Elles ne sortent PAS d'`EvolForm`, qui reste générique : l'écran les
	 *   fusionne dans la charge utile (`optionsVersTicket`). Elles voyagent ici
	 *   parce que c'est ce type que les relais recopient — et un relais qui
	 *   énumère ses champs jette ce qu'il ne nomme pas (#529). */
	epingle?: boolean;
	urgente?: boolean;
	confidentiel?: boolean;
}

/**
 * Le contenu riche est-il vide ? (balises retirées, espaces compris)
 *
 * Sorti d'`EvolForm` le 05/09/2026 avec les deux règles qui suivent : ce sont
 * des règles **du fil**, pas de l'affichage — elles répondent à « cette entrée
 * dit-elle quelque chose ? », question qui se pose pareil quel que soit l'écran.
 */
export function contenuRicheVide(html: string): boolean {
	return !html || html.replace(/<[^>]+>/g, '').trim() === '';
}

/**
 * 🔴 LE GESTE EST DÉDUIT, il ne se déclare pas.
 *
 * Une pastille laissée sur l'état courant ne change rien : l'entrée est un
 * commentaire. En choisir une autre en fait un changement d'état. C'est ce qui
 * permet UN seul point d'entrée à l'écran — la question « lequel des deux ? » a
 * déjà sa réponse dans ce que l'utilisateur a fait.
 *
 * Une CORRECTION (`editMode`) n'est jamais un changement d'état : on relit un
 * texte, on ne fait pas avancer le dossier.
 */
export function typeDeLEntree(
	editMode: boolean,
	nouveauStatut: string,
	statutCourant: string,
): 'commentaire' | 'etat' {
	return !editMode && nouveauStatut && nouveauStatut !== statutCourant ? 'etat' : 'commentaire';
}

/**
 * Une entrée vaut si elle APPORTE quelque chose : un changement d'état, un
 * texte, ou une pièce jointe. Rien des trois → rien à enregistrer.
 */
export function entreeEnregistrable(
	type: 'commentaire' | 'etat',
	contenu: string,
	nbFichiers: number,
): boolean {
	return type === 'etat' || !(contenuRicheVide(contenu) && nbFichiers === 0);
}

/**
 *  L'état INITIAL d'une entrée de fil, tel qu'`EvolForm` l'ouvre.
 *
 *  Sorti du formulaire le 17/09/2026 (modularité) : trois règles de DONNÉE qui
 *  n'ont rien d'un écran, et qui se lisent — et se testent — sans le monter :
 *    · le périmètre est celui que l'entrée avait déclaré en correction, sinon
 *      l'HÉRITÉ (`perimetreHerite`, 31/08/2026) ;
 *    · les destinataires se replient sur le défaut du site : une liste vide
 *      serait un effacement ;
 *    · 7. Photos · 8. Documents — DEUX sections, jamais une seule (cadre #430) ;
 *      le tri est une règle de FICHIERS (`separerFichiers`), pas de formulaire.
 */
export function etatInitialEntree(
	editMode: boolean,
	initialPerimetre: string[],
	perimetreCourant: string[],
	entrees: { perimetre_cible?: string[] | null }[],
	initialDestinataires: string[],
	initialFichiers: { url: string }[],
): { perimetre: string[]; destinataires: string[]; photos: string[]; documents: string[] } {
	const perimetre =
		editMode && initialPerimetre.length
			? [...initialPerimetre]
			: perimetreHerite(perimetreCourant, entrees);
	const destinataires = initialDestinataires.length ? [...initialDestinataires] : ['résidents'];
	const tries = separerFichiers(editMode ? initialFichiers.map((f) => f.url) : []);
	return { perimetre, destinataires, photos: tries.photos, documents: tries.documents };
}

/**
 * Les sections qu'un ÉCRAN HÔTE pose dans une Suite — la Catégorie, et ce qu'une
 * Suite d'affaire y ajoute : Équipement · Quand · Intervenant pour le conseil
 * (`SectionsSuiteConseil`, #1207), et la Mise en avant (`OptionsEvolutionTicket`,
 * `SectionOptionsPublication`).
 *
 * 🔴 Le créneau ne s'ouvrait que sur `nature` (25/09/2026). L'affaire la déclare
 * `hérité` en évolution : sa Suite ne rendait donc AUCUNE de ces sections, et une
 * affaire ne pouvait pas devenir urgente en cours de suivi.
 */
const SECTIONS_DE_L_HOTE: readonly IdSection[] = [
	'nature',
	'equipement',
	'quand',
	'intervenant',
	'mise_en_avant',
];

/**
 * Les TROIS créneaux d'`EvolForm`, à leur rang (#1326, 25/09/2026).
 *
 * Il n'y en avait qu'UN, rendu après le Suivi et avant le Périmètre : l'Équipement
 * (rang 2) y passait APRÈS le Suivi (3), et la Mise en avant (12) avant le
 * Périmètre (7). Signalé à l'écran : « l'ordre de section d'une affaire n'est pas
 * le même en mode édition et en mode suite ».
 *
 * Le créneau d'une section se DÉDUIT de `SECTIONS_ORDRE` — avant le Suivi, avant
 * le Périmètre, ou après les Destinataires — jamais d'une seconde table : c'est
 * elle qui avait laissé la Suite dans l'ordre d'avant les treize sections.
 */
export type Creneau = 'avant_suivi' | 'specifiques' | 'mise_en_avant';

export function creneauDe(id: IdSection): Creneau {
	const rang = SECTIONS_ORDRE.indexOf(id);
	if (rang < SECTIONS_ORDRE.indexOf('suivi')) return 'avant_suivi';
	if (rang < SECTIONS_ORDRE.indexOf('perimetre')) return 'specifiques';
	return 'mise_en_avant';
}

/** Le créneau s'ouvre dès que l'une de ses sections est déclarée en évolution. */
function creneauPresent(entite: EntiteDeclaree, creneau: Creneau): boolean {
	return SECTIONS_DE_L_HOTE.some(
		(id) => creneauDe(id) === creneau && sectionPresente(entite, 'evolution', id),
	);
}

/**
 * Une section du CIBLAGE est-elle offerte dans la Suite, pour cet objet-là ?
 *
 * Présente dans l'état `evolution` ET non éteinte par ce que l'objet EST
 * (`conditions` — la nature d'une affaire, `natureDe`). Une section éteinte
 * ne se rend pas du tout dans une Suite : dans le formulaire, elle reste grisée
 * pour montrer ce qu'un changement de catégorie rallumerait ; une Suite ne
 * change pas la catégorie, il n'y a donc rien à rallumer.
 *
 * 🔴 Les Destinataires d'une affaire SUIVIE ne s'offrent pas dans la Suite
 * (25/09/2026) — voir `sectionsDeLaSuite`.
 */
function sectionDeLaSuite(
	entite: EntiteDeclaree,
	id: IdSection,
	conditions: readonly ConditionInactive[],
): boolean {
	return (
		sectionPresente(entite, 'evolution', id) && !motifInactif(entite, 'evolution', id, conditions)
	);
}

/** Les sections qu'une Suite offre — `EvolForm` n'en décide plus lui-même. */
export interface SectionsDeLaSuite {
	perimetre: boolean;
	destinataires: boolean;
	avantSuivi: boolean;
	specifiques: boolean;
	miseEnAvant: boolean;
	piecesJointes: boolean;
	diffusion: boolean;
}

/**
 * Ce qu'une Suite (ou la correction d'une entrée) offre, pour CET objet et CE
 * lecteur. Chaque ligne combine ce qui EXISTE (la déclaration de l'entité) et
 * ce que cet utilisateur-ci PEUT (le droit, passé par l'appelant) — jamais l'un
 * à la place de l'autre (#463). Extrait d'`EvolForm` le 25/09/2026 : la règle
 * est pure, et le composant dépassait 500 lignes.
 *
 * 🔴 LE PÉRIMÈTRE SE CORRIGE AUSSI (01/09/2026, à l'écran) :
 *
 * > *« L'édition peut modifier le périmètre (correction d'erreur
 * > d'affectation d'un périmètre) »*
 *
 * La ligne valait `avecPerimetre && !editMode`, au motif que « préciser est un
 * geste de SUIVI, qui raturerait un fait daté en réécrivant une entrée passée »
 * — et le serveur refusait le champ en PATCH, ce qui fermait la question. Le
 * motif vaut pour un RESSERREMENT, pas pour une faute de clic. Et la faute coûte
 * cher : le périmètre d'une entrée écrase celui du ticket, donc une erreur
 * d'affectation reclasse tout le ticket.
 *
 * ⚠️ Côté serveur, la correction ne se propage à l'objet que si l'entrée
 * corrigée est la dernière à avoir précisé quelque chose
 * (`app/utils/perimetre_fil.py`) : corriger une vieille entrée ne défait pas une
 * précision récente.
 */
export function sectionsDeLaSuite(
	entite: EntiteDeclaree,
	conditions: readonly ConditionInactive[],
	droits: {
		perimetre: boolean;
		piecesJointes: boolean;
		diffusion: boolean;
		/** Les créneaux que l'écran hôte REMPLIT (`$$slots`). */
		creneaux: Record<Creneau, boolean>;
	},
): SectionsDeLaSuite {
	//  🔴 Pas de Destinataires dans la Suite d'une affaire SUIVIE (25/09/2026).
	//  Le formulaire les lui ouvre depuis la v2.50.0 (#1296), mais pour une seule
	//  chose : la case « Confidentielle » — « pas de profils à choisir pour une
	//  affaire suivie » (`SectionDestinataires`). La Suite ne porte pas cette case,
	//  et elle proposait les profils : pour le conseil, le serveur les ÉCRIVAIT
	//  sur l'affaire, valeur qu'aucun écran ne montre et que la correction efface
	//  (`chargeUtileAffaire`). Il ne l'écrit plus (`add_evolution`).
	const affaireSuivie = conditions.includes('suivie');
	return {
		perimetre: droits.perimetre && sectionPresente(entite, 'evolution', 'perimetre'),
		destinataires: sectionDeLaSuite(entite, 'destinataires', conditions) && !affaireSuivie,
		avantSuivi: droits.creneaux.avant_suivi && creneauPresent(entite, 'avant_suivi'),
		specifiques: droits.creneaux.specifiques && creneauPresent(entite, 'specifiques'),
		miseEnAvant: droits.creneaux.mise_en_avant && creneauPresent(entite, 'mise_en_avant'),
		piecesJointes: droits.piecesJointes && sectionPresente(entite, 'evolution', 'pieces_jointes'),
		diffusion: droits.diffusion && sectionPresente(entite, 'evolution', 'diffusion'),
	};
}

File: api/app/utils/evolutions.py
"""Retirer une entrée d'un fil de suivi — **le geste, écrit une seule fois** (#512).

## Pourquoi ce module

Trois objets du site portent un fil d'évolutions : le **ticket**
(`TicketEvolution`), l'**actualité** (`PublicationEvolution`) et l'**événement**
(`EvenementEvolution`). Le geste de suppression n'existait que pour le premier.

Le recopier deux fois aurait été le réflexe évident — et exactement la faute que
`standards/02` interdit. Ce n'est pas une question de lignes : les trois copies
auraient porté chacune leur liste de types effaçables et leur message d'erreur,
et la première divergence serait passée inaperçue, parce qu'un administrateur ne
compare pas le refus d'un écran à celui d'un autre.

⚠️ Le ticket #512 le disait déjà : *« les nouveaux endpoints doivent dire **la
même chose** que lui, ni plus ni moins »*. La seule façon de le garantir est
qu'ils appellent le même code.

## Les deux arbitrages, tranchés le 18/08/2026 et NON rouverts

**Qui** — `require_admin`, jamais `require_cs_or_admin`. Corriger sa propre
entrée (`PATCH`) est un geste ordinaire ; **effacer** fait disparaître une trace
que d'autres ont pu lire et sur laquelle ils ont pu agir. C'est la frontière de
« archiver n'est pas supprimer » : l'irréversible reste à l'administrateur. La
dépendance est posée par chaque endpoint, pas ici — c'est le routeur qui déclare
qui il laisse entrer.

**Quoi** — `commentaire` et `etat` seulement. Une **réponse** appartient à son
auteur, souvent un résident : l'effacer supprimerait la parole de quelqu'un
d'autre, ce qui n'est pas la même chose que retirer une ligne écrite par le
système ou par soi-même.

⚠️ Supprimer une entrée d'état **ne change pas** le statut de l'objet : celui-ci
vit dans sa propre colonne, le fil n'en est que le récit. Le coût est une perte
de traçabilité — on ne saura plus quand l'objet est passé « En cours » —, pas une
incohérence de données. C'est un coût réel : le fil sert de preuve au conseil
syndical face au syndic, et une transition effacée ne se retrouve pas.
"""

from typing import Any

from fastapi import HTTPException
from sqlmodel import Session

from app.auth.deps import peut_editer

#: Les types d'entrée qu'un humain ÉCRIT — par opposition à celles que le serveur
#: trace lui-même (« Correction : … »).
#:
#: 🔴 C'est LE fait dont dépendent les trois gestes du fil, et il était écrit
#: QUATRE fois (14/09/2026, #779) : ici, dans les deux `PATCH` et dans la
#: création d'une entrée de ticket. Créer, corriger et effacer posent la même
#: question — *cette entrée porte-t-elle le texte de quelqu'un ?* — et une entrée
#: tracée automatiquement répond non aux trois.
#:
#: ⚠️ Le jour où les trois divergeraient — un type créable mais non corrigeable,
#: par exemple — la seconde liste s'écrirait ICI, avec son motif, et non dans une
#: route.
TYPES_SAISIS: tuple[str, ...] = ("commentaire", "etat")

#: Le nom sous lequel la suppression connaît cette liste, et sous lequel le FRONT
#: la tient (`RubriqueHistorique.svelte`).
#:
#: ⚠️ Conservé tel quel : les contextes de build sont `./api` et `./front`, rien
#: de la racine n'entre dans les images — seule la copie est possible, et
#: `test_evolutions_suppression.py` échoue si les deux divergent. Renommer des deux
#: côtés est un lot à part ; le faire en passant casserait le seul lien qui les
#: tient.
TYPES_EFFACABLES: tuple[str, ...] = TYPES_SAISIS

#: Un seul message pour les trois fils. Deux formulations différentes pour le
#: même refus laisseraient croire à deux règles différentes.
REFUS_TYPE = "Cette entrée ne peut pas être supprimée : une réponse appartient à son auteur."


def evolution_modifiable(
    session: Session,
    modele: type,
    evol_id: int,
    *,
    champ_parent: str,
    parent_id: int,
    user,
) -> Any:
    """L'entrée, si elle peut être CORRIGÉE par qui la demande — sinon l'erreur.

    🔴 Cette garde était écrite DEUX fois (14/09/2026, #779), dans les deux
    routes `PATCH …/evolutions/{id}` — publications et tickets —, y compris le
    tuple `("commentaire", "etat")` **recopié à la main à deux pas du module qui
    le porte déjà**. Trois copies d'une même liste, dont deux qui s'ignoraient.

    ⚠️ **Pourquoi le même tuple que la suppression**, alors que ce sont deux
    gestes : les deux répondent à la même question de fait — *cette entrée
    porte-t-elle un texte écrit par quelqu'un ?* Une entrée « Correction : … »
    tracée automatiquement ne décrit le geste de personne : on ne la corrige pas
    plus qu'on ne l'efface.

    Le jour où les deux divergeraient — un type modifiable mais non effaçable —
    c'est ICI que la seconde liste s'écrirait, avec son motif, et non dans une
    route.

    ⚠️ Les trois refus dans cet ordre, et il compte : l'appartenance d'abord
    (404 — ne pas révéler l'existence), le type ensuite (422 — la demande est
    recevable mais l'objet s'y refuse), le droit en dernier (403). L'inverse
    dirait « accès refusé » sur une entrée qui n'existe pas.
    """
    evol: Any = session.get(modele, evol_id)
    if not evol or getattr(evol, champ_parent, None) != parent_id:
        raise HTTPException(404, "Évolution introuvable")
    if getattr(evol, "type", None) not in TYPES_SAISIS:
        raise HTTPException(422, "Ce type d'évolution ne peut pas être modifié")
    #  L'auteur ou un admin — `peut_editer`, du module central d'autorisation.
    if not peut_editer(evol, user):
        raise HTTPException(403, "Accès refusé")
    return evol


def supprimer_evolution(
    session: Session,
    modele: type,
    evol_id: int,
    *,
    champ_parent: str,
    parent_id: int,
) -> None:
    """Retire une entrée de fil, ou lève l'exception qui explique pourquoi non.

    `champ_parent` est le nom de la clé étrangère (`ticket_id`,
    `publication_id`, `evenement_id`) : c'est la seule chose qui distingue les
    trois fils. La vérifier n'est pas une formalité — sans elle, un identifiant
    d'entrée valide permettrait d'effacer l'entrée d'un AUTRE objet que celui
    dont on a l'adresse, et le contrôle d'accès de l'URL ne servirait à rien.
    """
    evol: Any = session.get(modele, evol_id)
    if not evol or getattr(evol, champ_parent, None) != parent_id:
        raise HTTPException(404, "Entrée introuvable")
    if getattr(evol, "type", None) not in TYPES_EFFACABLES:
        raise HTTPException(422, REFUS_TYPE)
    session.delete(evol)
    session.commit()
